nrf_security/src/mbedtls/shared/CMakeLists.txt

```
@@ -36,25 +20,54 @@ add_library(mbedcrypto_shared STATIC
   ${ZEPHYR_BASE}/misc/empty_file.c
 )
 
-target_sources_ifdef(CONFIG_VANILLA_MBEDTLS_CHACHA20_C
-  mbedcrypto_shared PRIVATE ${ARM_MBEDTLS_PATH}/library/chacha20.c
-)
-target_sources_ifdef(CONFIG_VANILLA_MBEDTLS_POLY1305_C
-  mbedcrypto_shared PRIVATE ${ARM_MBEDTLS_PATH}/library/poly1305.c
-)
-target_sources_ifdef(CONFIG_VANILLA_MBEDTLS_CHACHAPOLY_C
-  mbedcrypto_shared PRIVATE ${ARM_MBEDTLS_PATH}/library/chachapoly.c
-)
-target_sources_ifdef(CONFIG_VANILLA_MBEDTLS_RSA_C
-  mbedcrypto_shared PRIVATE ${ARM_MBEDTLS_PATH}/library/rsa.c
-)
-target_sources_ifdef(CONFIG_VANILLA_MBEDTLS_RSA_C
-  mbedcrypto_shared PRIVATE ${ARM_MBEDTLS_PATH}/library/rsa_internal.c
-)
+if (CONFIG_VANILLA_MBEDTLS_CHACHA20_C OR
+  (CONFIG_MBEDTLS_CHACHA20_C AND NOT CONFIG_CC310_MBEDTLS_CHACHA20_C))
+
+  target_sources(mbedcrypto_shared PRIVATE
+    ${ARM_MBEDTLS_PATH}/library/chacha20.c
+  )
+endif()
+
+if (CONFIG_VANILLA_MBEDTLS_POLY1305_C OR
+  (CONFIG_MBEDTLS_POLY1305_C AND NOT CONFIG_CC310_MBEDTLS_POLY1305_C))
+
+  target_sources(mbedcrypto_shared PRIVATE
+    ${ARM_MBEDTLS_PATH}/library/poly1305.c
+  )
+endif()
+
+if (CONFIG_VANILLA_MBEDTLS_CHACHAPOLY_C OR
+  (CONFIG_MBEDTLS_CHACHAPOLY_C AND NOT CONFIG_CC310_MBEDTLS_CHACHAPOLY_C))
+
+  target_sources(mbedcrypto_shared PRIVATE
+    ${ARM_MBEDTLS_PATH}/library/chachapoly.c
+  )
+endif()
+
+if (CONFIG_VANILLA_MBEDTLS_RSA_C OR
+   (CONFIG_MBEDTLS_RSA_C AND NOT CONFIG_CC310_MBEDTLS_RSA_C))
```

If you have single backend CC310 and enable `MBEDTLS_RSA_C` then those files gets added, because `CONFIG_CC310_MBEDTLS_RSA_C` is only true if there are multiple backends.
https://github.com/nrfconnect/sdk-nrfxlib/blob/master/nrf_security/Kconfig#L973-L1003

Is that intended ?

Or would this be more correct:
```suggestion
if (CONFIG_VANILLA_MBEDTLS_RSA_C OR
   (CONFIG_MBEDTLS_RSA_C AND NOT CONFIG_CC310_BACKEND))
```

---

Unfortunately it is a little bit deeper than that as there is cmake-logic as well. I will keep it as-is for now but this (and other related topics) will be handled in the issue NCSDK-6047

...in which more logic is moved from cmake to Kconfig. One of the effects of this work is that  CONFIG_CC310_MBEDTLS_RSA_C will be set on CC310 as single-backend directly in Kconfig